File: examples/user_distance.c
/* cell resampling with custom distance using C
 *
 * see `cres.c` for instructions on running examples
 */

#include "cres.h"

#include <math.h>
#include <stdio.h>

/* user-defined distance function
 *
 * note that this function has be thread-safe and must never return NaN!
 *
 * this function is just for demonstration
 * and doesn't make much sense physically
 */
double my_distance(
  void* data,              /* user-defined data */
  EventView const * ev1,
  EventView const * ev2
) {
  /* `data` can be used to pass our own parameters */
  const double E_fact = * (const double*) data;

  double dist = 0.;

  /* iterate over particle types
   *
   * for simplicity, we only compare events that have the same particle types
   * and the same number of particles for each type
   */
  if(ev1->n_type_sets != ev2->n_type_sets) return INFINITY;
  for(uintptr_t t = 0; t < ev1->n_type_sets; ++t) {
    if(
      (ev1->type_sets[t].pid != ev2->type_sets[t].pid)
      || (ev1->type_sets[t].n_momenta != ev2->type_sets[t].n_momenta)
    ) return INFINITY;

    for(uintptr_t i = 0; i < ev1->type_sets[t].n_momenta; ++i) {
      /* use d(p1,p2) = E_fact * |E1 - E2| + |p1_x - p2_x| */
      const double* p1 = ev1->type_sets[t].momenta[i];
      const double* p2 = ev2->type_sets[t].momenta[i];
      dist += E_fact * fabs(p1[0] - p2[0]) + fabs(p1[1] - p2[1]);
    }
  }
  return dist;
}

int main(int argc, char** argv) {
  if(argc < 3) {
    return 1;
  }

  int32_t res = cres_logger_from_env("CRES_LOG");
  if(res != 0) cres_print_last_err();

  Opt opt;
  opt.n_infiles = argc - 2;
  opt.infiles = argv + 1;
  opt.outfile = argv[argc - 1];

  opt.jet_def.algorithm = AntiKt;
  opt.jet_def.radius = 0.4;
  opt.jet_def.min_pt = 30.;
  opt.neighbour_search = Tree;
  opt.max_cell_size = INFINITY;

  /* custom distance function with `E_fact` as extra data */
  double E_fact = 0.5;
  DistanceFn dist = {
    .fun = my_distance,
    .data = &E_fact
  };
  opt.distance = &dist;

  /* build and run the resampler */
  res = cres_run(&opt);
  if(res != 0) cres_print_last_err();
  return res;
}

File: examples/user_distance.rs
// cell resampling with user-defined distance
// run with `cargo run --release --example user_distance -- IN.hepmc OUT.hepmc`
// set the environment variable `RUST_LOG=info` for command-line output
use std::error::Error;

use cres::distance::Distance;
use cres::event::Event;
use cres::prelude::*;

use env_logger;
use noisy_float::prelude::*;

// this distance is just for demonstration
// and doesn't make much sense physically
struct MyDistance {
    e_fact: N64,
}

impl Distance for MyDistance {
    fn distance(&self, ev1: &Event, ev2: &Event) -> N64 {
        if ev1.outgoing().len() != ev2.outgoing().len() {
            return N64::infinity();
        }
        let mut dist = n64(0.);
        let set_pairs = ev1.outgoing().iter().zip(ev2.outgoing().iter());
        for ((id1, s1), (id2, s2)) in set_pairs {
            if id1 != id2 || s1.len() != s2.len() {
                return N64::infinity();
            }
            dist += s1
                .iter()
                .zip(s2.iter())
                .map(|(p1, p2)| {
                    self.e_fact * (p1[0] - p2[0]).abs() + (p1[1] - p2[1]).abs()
                })
                .sum::<N64>();
        }
        n64(0.)
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    env_logger::init();

    let mut args = std::env::args();
    let _ = args.next().unwrap(); // ignore program name
    let infile = args.next().unwrap().into();
    let outfile = args.next().unwrap().into();

    let event_io = IOBuilder::default();
    let event_io = event_io.build_from_files(infile, outfile)?;

    let converter = Converter::new();

    let resampler = ResamplerBuilder::default()
        .distance(MyDistance { e_fact: n64(0.5) })
        .build();

    let mut cres = CresBuilder {
        event_io,
        converter,
        clustering: NO_CLUSTERING, // disable jet clustering
        resampler,
        unweighter: NO_UNWEIGHTING, // disable unweighting
    }
    .build();
    cres.run()?;
    Ok(())
}
